#### Carnegie Mellon University - Tepper School of Business
#### 46-881 - Programming in R and Python
#### 22/23 - Mini 2 - Sections MO2 - Farahat

### FINAL EXAM 

**Due December 18th at 11:59 p.m. Pittsburgh time.**

---

Please provide your name and Andrew ID below:

- Name (Last, First):Hrishikesh Mysore Harishkumar

- Andrew ID: hmysoreh

---

### PROBLEM 1 (55 points)

The CDC Covid Tracker website includes a section on Health Equity Data ([link](https://covid.cdc.gov/covid-data-tracker/#health-equity-data)) with the following statement:

> Health equity is when everyone has the opportunity to be as healthy as possible. Achieving health equity requires valuing everyone equally with focused and ongoing efforts to address avoidable inequities, historical and contemporary injustices, and the elimination of disparities in health and healthcare. The population health impact of COVID-19 has exposed longstanding inequities that have systematically undermined the physical, social, economic, and emotional health of racial and ethnic minority populations and other population groups that are bearing a disproportionate burden of COVID-19.

In this problem, I ask you to use Python (pandas) to analyze data supplied by the CDC on Covid-19 vaccination rates along with variables measuring the degree of social vulnerability in US counties (see [this link](https://www.atsdr.cdc.gov/placeandhealth/svi/documentation/SVI_documentation_2018.html) for definitions). More information on the data and its context is available on the CDC website. 

The dataset, <vaccination_equity.csv>, is available on Canvas accompanying this problem. A data dictionary is provided in the table below. (Note: I've downloaded this dataset a little over a year ago, on October 14th 2021, so it's a bit out of date especially with respect to vaccination rates.) 

| Column name |	Description |
| :-----------| :------------------|
| STATE	      | State abbreviation |
| COUNTY	  | County name        |
| FIPS	      | Unique county code |
| E_TOTPOP	  | Population estimate |
| EP_POV      |	Percentage of persons below poverty estimate|
| EP_UNEMP    |	Unemployment Rate estimate|
| EP_PCI      |	Per capita income estimate|
| EP_NOHSDP   |	Percentage of persons with no high school diploma (age 25+) estimate|
| EP_AGE65    |	Percentage of persons aged 65 and older estimate|
| EP_AGE17    |	Percentage of persons aged 17 and younger estimate|
| EP_DISABL   |	Percentage of civilian noninstitutionalized population with a disability estimate|
| EP_SNGPNT   |	Percentage of single parent households with children under 18 estimate|
| EP_MINRTY   |	Percentage minority (all persons except white, non-Hispanic) estimate|
| EP_LIMENG   |	Percentage of persons (age 5+) who speak English "less than well" estimate|
| EP_MUNIT    |	Percentage of housing in structures with 10 or more units estimate|
| EP_MOBILE   |	Percentage of mobile homes estimate|
| EP_CROWD    |	Percentage of occupied housing units with more people than rooms estimate|
| EP_NOVEH    |	Percentage of households with no vehicle available estimate|
| EP_GROUPQ   |	Percentage of persons in group quarters estimate|
| SVI         | County Social Vulnerability Index on a catgeorical scale|
| V_REPORT_COMPLETENESS | Percent of reporting completeness within county| 
| V_PCT_ALL   |	Percent of total population fully vaccinated|
| V_PCT_12YRS |	Percent of population 12 years and older fully vaccinated|
| V_PCT_18YRS |	Percent of population 18 years and older fully vaccinated|
| V_PCT_65YRS |	Percent of population 65 years and older fully vaccinated| 


**Answer the following questions based on the data provided**. 

#### Part (a)

How many counties are represented in this dataset? 

In [1]:
import pandas as pd

In [2]:
df=pd.read_csv("vaccination_equity.csv")
df

,STATE,COUNTY,FIPS,E_TOTPOP,EP_POV,EP_UNEMP,EP_PCI,EP_NOHSDP,EP_AGE65,EP_AGE17,...,EP_MOBILE,EP_CROWD,EP_NOVEH,EP_GROUPQ,SVI,V_REPORT_COMPLETENESS,V_PCT_ALL,V_PCT_12YRS,V_PCT_18YRS,V_PCT_65YRS
0,AL,Autauga,1001,55200,15.4,4.2,29372,11.3,14.6,24.2,...,18.4,1.4,5.6,1.0,Low-Mod,92.5,34.7,40.8,42.5,62.2
1,AL,Blount,1009,57645,14.4,4.1,22656,19.8,17.8,23.4,...,25.2,1.6,4.2,0.9,Low-Mod,92.5,27.2,32.0,33.9,50.1
2,AL,Butler,1013,20025,23.5,6.7,20430,15.4,19.0,22.8,...,26.2,1.8,7.8,1.6,High,92.5,34.4,40.1,42.1,59.6
3,AL,Calhoun,1015,115098,18.6,8.8,24706,15.9,16.8,21.9,...,14.7,1.7,5.8,2.7,High,92.5,40.3,46.8,49.2,74.6
4,AL,Chambers,1017,33826,16.6,5.0,22827,18.6,18.9,20.7,...,14.0,3.0,7.3,1.5,Mod-High,92.5,27.3,31.7,33.2,48.8
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3136,TX,Duval,48131,11355,25.6,10.6,17864,32.9,17.8,26.1,...,13.0,7.4,10.0,4.4,High,NaN,NaN,NaN,NaN,NaN
3137,TX,Zapata,48505,14369,39.5,11.0,17228,40.0,13.9,33.6,...,34.2,18.1,7.9,0.2,High,NaN,NaN,NaN,NaN,NaN
3138,TX,Zavala,48507,12131,34.8,8.4,13350,38.0,13.7,29.5,...,22.8,5.6,14.4,1.6,High,NaN,NaN,NaN,NaN,NaN
3139,TX,Dimmit,48127,10663,29.8,16.4,18121,32.7,16.3,30.0,...,16.7,8.5,11.9,1.2,High,NaN,NaN,NaN,NaN,NaN


In [3]:
df.shape

(3141, 25)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3141 entries, 0 to 3140
Data columns (total 25 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   STATE                  3141 non-null   object 
 1   COUNTY                 3141 non-null   object 
 2   FIPS                   3141 non-null   int64  
 3   E_TOTPOP               3141 non-null   int64  
 4   EP_POV                 3141 non-null   float64
 5   EP_UNEMP               3141 non-null   float64
 6   EP_PCI                 3141 non-null   int64  
 7   EP_NOHSDP              3141 non-null   float64
 8   EP_AGE65               3141 non-null   float64
 9   EP_AGE17               3141 non-null   float64
 10  EP_DISABL              3141 non-null   float64
 11  EP_SNGPNT              3141 non-null   float64
 12  EP_MINRTY              3141 non-null   float64
 13  EP_LIMENG              3141 non-null   float64
 14  EP_MUNIT               3141 non-null   float64
 15  EP_M

In [5]:
len(df["COUNTY"])

3141

#### ANSWER: A total of 3141 counties are represented in the dataset.

#### Part (b)

What is the total population of the counties represented in the dataset? 

In [6]:
sum(df["E_TOTPOP"])

322863723

#### ANSWER: The total population of the counties represented in 322863723

#### Part (c)

How many counties are missing a value in at least one of the columns?

In [7]:
df[df.isna().any(axis=1) == True]

,STATE,COUNTY,FIPS,E_TOTPOP,EP_POV,EP_UNEMP,EP_PCI,EP_NOHSDP,EP_AGE65,EP_AGE17,...,EP_MOBILE,EP_CROWD,EP_NOVEH,EP_GROUPQ,SVI,V_REPORT_COMPLETENESS,V_PCT_ALL,V_PCT_12YRS,V_PCT_18YRS,V_PCT_65YRS
27,AK,Valdez-Cordova,2261,9301,8.1,6.6,38507,4.0,12.4,23.6,...,11.3,2.3,7.1,2.7,NaN,NaN,NaN,NaN,NaN,NaN
50,CA,Inyo,6027,18085,10.2,5.2,31540,11.2,22.9,20.1,...,21.8,1.9,6.8,3.6,Mod-High,NaN,NaN,NaN,NaN,NaN
112,ID,Ada,16001,446052,11.0,3.8,33352,4.9,13.5,24.5,...,4.0,1.4,3.8,1.9,Low,97.6,51.0,NaN,66.3,91.0
113,ID,Bonner,16017,42711,14.7,3.7,27590,9.2,23.0,20.0,...,12.7,2.0,4.4,1.0,Low,97.6,36.3,NaN,45.1,66.8
114,ID,Boundary,16021,11549,16.8,3.9,25401,8.0,21.1,23.7,...,10.9,2.9,2.6,1.0,Low-Mod,97.6,28.9,NaN,37.5,59.8
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3136,TX,Duval,48131,11355,25.6,10.6,17864,32.9,17.8,26.1,...,13.0,7.4,10.0,4.4,High,NaN,NaN,NaN,NaN,NaN
3137,TX,Zapata,48505,14369,39.5,11.0,17228,40.0,13.9,33.6,...,34.2,18.1,7.9,0.2,High,NaN,NaN,NaN,NaN,NaN
3138,TX,Zavala,48507,12131,34.8,8.4,13350,38.0,13.7,29.5,...,22.8,5.6,14.4,1.6,High,NaN,NaN,NaN,NaN,NaN
3139,TX,Dimmit,48127,10663,29.8,16.4,18121,32.7,16.3,30.0,...,16.7,8.5,11.9,1.2,High,NaN,NaN,NaN,NaN,NaN


In [8]:
len(df[df.isna().any(axis=1) == True])

395

#### ANSWER: There are 395 counties that are missing a value in at least one of the columns

**For the remaining parts of this problem, delete all the counties that have missing values in at least one column.** This will possibly bias the data, but ignore that for the purposes of this exam. Remove only rows with missing (NA) values; do not consider V_Report_Completeness < 100% as a missing value.

In [9]:
df_copy = df.dropna()

In [23]:
len(df_copy)

2746

In [24]:
df_copy

,STATE,COUNTY,FIPS,E_TOTPOP,EP_POV,EP_UNEMP,EP_PCI,EP_NOHSDP,EP_AGE65,EP_AGE17,...,EP_MOBILE,EP_CROWD,EP_NOVEH,EP_GROUPQ,SVI,V_REPORT_COMPLETENESS,V_PCT_ALL,V_PCT_12YRS,V_PCT_18YRS,V_PCT_65YRS
0,AL,Autauga,1001,55200,15.4,4.2,29372,11.3,14.6,24.2,...,18.4,1.4,5.6,1.0,Low-Mod,92.5,34.7,40.8,42.5,62.2
1,AL,Blount,1009,57645,14.4,4.1,22656,19.8,17.8,23.4,...,25.2,1.6,4.2,0.9,Low-Mod,92.5,27.2,32.0,33.9,50.1
2,AL,Butler,1013,20025,23.5,6.7,20430,15.4,19.0,22.8,...,26.2,1.8,7.8,1.6,High,92.5,34.4,40.1,42.1,59.6
3,AL,Calhoun,1015,115098,18.6,8.8,24706,15.9,16.8,21.9,...,14.7,1.7,5.8,2.7,High,92.5,40.3,46.8,49.2,74.6
4,AL,Chambers,1017,33826,16.6,5.0,22827,18.6,18.9,20.7,...,14.0,3.0,7.3,1.5,Mod-High,92.5,27.3,31.7,33.2,48.8
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3127,SD,Todd,46121,10146,55.1,13.2,10931,22.1,7.5,40.3,...,17.8,13.5,21.6,0.4,High,84.8,45.6,64.0,67.5,81.7
3128,SD,Ziebach,46137,2814,41.0,23.1,14943,15.9,9.0,32.0,...,27.2,14.2,12.2,0.0,High,84.8,24.6,29.0,30.2,48.5
3133,MS,Claiborne,28021,9120,44.0,15.6,13503,22.6,15.3,21.8,...,37.8,5.7,10.4,8.0,High,98.3,47.0,54.6,54.5,77.8
3134,SC,Allendale,45005,9214,26.8,16.4,14744,22.3,18.4,19.0,...,32.9,2.1,20.3,12.7,High,93.8,50.5,57.5,59.8,73.0


In [25]:
df_copy[df_copy.isna().any(axis=1) == True]

,STATE,COUNTY,FIPS,E_TOTPOP,EP_POV,EP_UNEMP,EP_PCI,EP_NOHSDP,EP_AGE65,EP_AGE17,...,EP_MOBILE,EP_CROWD,EP_NOVEH,EP_GROUPQ,SVI,V_REPORT_COMPLETENESS,V_PCT_ALL,V_PCT_12YRS,V_PCT_18YRS,V_PCT_65YRS


#### Part (d)

Create a pivot table with a row for each State and a column for each Social Vulnerability Index category. Display in the table the population (summed across counties) of each State and SVI category combination.

In [26]:
df_copy.pivot_table(index = "STATE",  aggfunc="sum", values = ["E_TOTPOP","EP_POV","EP_UNEMP","EP_PCI","EP_NOHSDP","EP_AGE65","EP_AGE17","EP_DISABL","EP_SNGPNT","EP_MINRTY","EP_LIMENG","EP_MUNIT","EP_MOBILE","EP_CROWD","EP_NOVEH","EP_GROUPQ","SVI","V_REPORT_COMPLETENESS","V_PCT_ALL","V_PCT_12YRS","V_PCT_18YRS","V_PCT_65YRS"])



,EP_AGE17,EP_AGE65,EP_CROWD,EP_DISABL,EP_GROUPQ,EP_LIMENG,EP_MINRTY,EP_MOBILE,EP_MUNIT,EP_NOHSDP,...,EP_PCI,EP_POV,EP_SNGPNT,EP_UNEMP,E_TOTPOP,V_PCT_12YRS,V_PCT_18YRS,V_PCT_65YRS,V_PCT_ALL,V_REPORT_COMPLETENESS
STATE,,,,,,,,,,,,,,,,,,,,,
AK,234.2,111.3,137.4,128.2,62.6,18.1,587.9,46.2,35.7,111.1,...,255070,145.6,112.5,105.2,62842,586.1,592.4,682.8,483.8,869.4
AL,1480.5,1190.7,117.0,1272.7,175.5,54.7,2344.6,1430.1,214.1,1192.1,...,1545880,1362.6,602.4,538.5,4864680,2934.6,3076.4,4491.2,2519.1,6197.5
AR,1694.7,1426.5,197.4,1545.4,253.7,79.6,1847.1,1295.5,220.4,1213.7,...,1683205,1487.5,692.9,483.5,2990671,3465.0,3595.5,4850.2,2962.3,6765.0
AZ,347.2,301.1,80.1,239.0,54.0,70.2,743.2,324.2,69.5,239.2,...,351891,300.0,137.7,130.7,6946685,947.7,962.7,1156.7,811.9,1464.0
CA,1145.7,792.6,310.1,663.5,170.8,371.5,2452.0,339.5,446.3,824.2,...,1623539,760.5,455.1,366.5,39054386,3183.3,3256.9,3946.8,2721.7,4870.0
CO,1328.0,1169.5,158.3,881.2,238.4,151.0,1631.3,606.7,496.4,585.3,...,1964506,836.0,431.6,328.7,5531141,3614.6,3728.6,4708.2,3156.6,6233.6
CT,160.0,134.9,12.0,90.9,33.0,21.5,194.8,9.5,78.0,68.6,...,328170,73.3,64.6,48.1,3581504,610.9,620.8,736.6,537.0,787.2
DC,17.8,11.9,3.6,11.8,5.8,2.5,63.8,0.1,46.2,9.4,...,53321,16.8,8.4,7.4,684498,67.2,67.9,82.0,58.1,95.2
DE,64.0,57.1,5.4,39.6,7.3,5.8,104.7,30.7,26.2,33.7,...,97930,36.7,27.0,17.1,949495,189.7,194.6,251.3,163.8,288.6


#### Part (e)

Modify the pivot table created in Part (d) above so that the rows are scaled to sum to one for each State. Hint: For a pandas dataframe, say df, with numerical values (e.g. a pivot table), use df.div(df.sum(axis=1), axis=0) to scale the rows.

In [27]:
df_pivot = df_copy.pivot_table(index = "STATE",  aggfunc="sum", values = ["E_TOTPOP","EP_POV","EP_UNEMP","EP_PCI","EP_NOHSDP","EP_AGE65","EP_AGE17","EP_DISABL","EP_SNGPNT","EP_MINRTY","EP_LIMENG","EP_MUNIT","EP_MOBILE","EP_CROWD","EP_NOVEH","EP_GROUPQ","SVI","V_REPORT_COMPLETENESS","V_PCT_ALL","V_PCT_12YRS","V_PCT_18YRS","V_PCT_65YRS"])


In [28]:
df_pivot.div(df_pivot.sum(axis=1), axis=0)

,EP_AGE17,EP_AGE65,EP_CROWD,EP_DISABL,EP_GROUPQ,EP_LIMENG,EP_MINRTY,EP_MOBILE,EP_MUNIT,EP_NOHSDP,...,EP_PCI,EP_POV,EP_SNGPNT,EP_UNEMP,E_TOTPOP,V_PCT_12YRS,V_PCT_18YRS,V_PCT_65YRS,V_PCT_ALL,V_REPORT_COMPLETENESS
STATE,,,,,,,,,,,,,,,,,,,,,
AK,0.000725,0.000344,0.000425,0.000397,0.000194,0.000056,0.001819,1.429259e-04,0.000110,0.000344,...,0.789093,0.000450,0.000348,0.000325,0.194410,0.001813,0.001833,0.002112,0.001497,0.002690
AL,0.000230,0.000185,0.000018,0.000198,0.000027,0.000008,0.000364,2.219882e-04,0.000033,0.000185,...,0.239960,0.000212,0.000094,0.000084,0.755123,0.000456,0.000478,0.000697,0.000391,0.000962
AR,0.000360,0.000303,0.000042,0.000328,0.000054,0.000017,0.000392,2.751417e-04,0.000047,0.000258,...,0.357484,0.000316,0.000147,0.000103,0.635167,0.000736,0.000764,0.001030,0.000629,0.001437
AZ,0.000048,0.000041,0.000011,0.000033,0.000007,0.000010,0.000102,4.436812e-05,0.000010,0.000033,...,0.048158,0.000041,0.000019,0.000018,0.950683,0.000130,0.000132,0.000158,0.000111,0.000200
CA,0.000028,0.000019,0.000008,0.000016,0.000004,0.000009,0.000060,8.340433e-06,0.000011,0.000020,...,0.039885,0.000019,0.000011,0.000009,0.959442,0.000078,0.000080,0.000097,0.000067,0.000120
CO,0.000176,0.000155,0.000021,0.000117,0.000032,0.000020,0.000217,8.061139e-05,0.000066,0.000078,...,0.261021,0.000111,0.000057,0.000044,0.734915,0.000480,0.000495,0.000626,0.000419,0.000828
CT,0.000041,0.000034,0.000003,0.000023,0.000008,0.000005,0.000050,2.427176e-06,0.000020,0.000018,...,0.083845,0.000019,0.000017,0.000012,0.915046,0.000156,0.000159,0.000188,0.000137,0.000201
DC,0.000024,0.000016,0.000005,0.000016,0.000008,0.000003,0.000086,1.354224e-07,0.000063,0.000013,...,0.072209,0.000023,0.000011,0.000010,0.926964,0.000091,0.000092,0.000111,0.000079,0.000129
DE,0.000061,0.000054,0.000005,0.000038,0.000007,0.000006,0.000100,2.926637e-05,0.000025,0.000032,...,0.093357,0.000035,0.000026,0.000016,0.905156,0.000181,0.000186,0.000240,0.000156,0.000275


#### Part (f)

What is the total number of persons fully vaccinated in the counties represented in the dataset? 

In [29]:
sum(df_copy["E_TOTPOP"] * (df_copy["V_PCT_ALL"]/100))

153178737.36499995

#### ANSWER: The total number of persons fully vaccinated in the counties is equal to 153178737.4

#### Part (g)

Without adjusting for population, what is the median, across all counties, of V_PCT_12YRS (Percent of population 12 years and older fully vaccinated)? 

In [30]:
df_copy["V_PCT_12YRS"].median()

50.2

#### ANSWER: The median, across all counties, of V_PCT_12YRS is 50.2

#### Part (h)

Create a table showing the median of V_PCT_12YRS grouped by the Social Vulnerability Index, SVI.

In [31]:
df_copy.pivot_table(index = "SVI", aggfunc = "median", values = ["V_PCT_12YRS"])

,V_PCT_12YRS
SVI,
High,46.50
Low,53.90
Low-Mod,51.25
Mod-High,48.20


#### Part (i)
What is the coefficient of correlation between V_PCT_12YRS and EP_POV? Hint: Use the corr() method; see help(pd.DataFrame.corr).

In [32]:
print(df_copy["EP_POV"].corr(df_copy["V_PCT_12YRS"]))

-0.2522880086714833


#### ANSWER: The coefficient of correlation between V_PCT_12YRS and EP_POV is -0.2523

**The remaining parts focus on the state of Pennsylvania**.

#### Part (j)

Without adjusting for population, what is the median of V_PCT_12YRS across all counties in Pennsylvania?

In [33]:
df_copy["STATE"].unique()

array(['AL', 'AZ', 'AR', 'CA', 'CO', 'CT', 'DE', 'FL', 'GA', 'IL', 'IN',
       'IA', 'KS', 'KY', 'ME', 'MD', 'MA', 'MI', 'MN', 'MS', 'MO', 'MT',
       'NE', 'NV', 'NH', 'NJ', 'NM', 'NY', 'NC', 'ND', 'OH', 'OK', 'OR',
       'PA', 'RI', 'SC', 'SD', 'TN', 'UT', 'VT', 'VA', 'WA', 'WV', 'WI',
       'WY', 'AK', 'DC'], dtype=object)

In [34]:
df_copy[df_copy["STATE"] == 'PA'].pivot_table(index = "STATE", values ="V_PCT_12YRS", aggfunc = "median")

,V_PCT_12YRS
STATE,
PA,59.2


#### ANSWER: From the above table, the median of V_PCT_12YRS across all counties in Pennsylvania is 59.2

#### Part (k)

Which 5 counties in PA has the highest V_PCT_12YRS? Which 5 counties had the lowest?

In [35]:
df_pa_order = df_copy[df_copy["STATE"] == 'PA'].sort_values(by=["V_PCT_12YRS"], ascending=False)

In [36]:
df_pa_order.head(5)

,STATE,COUNTY,FIPS,E_TOTPOP,EP_POV,EP_UNEMP,EP_PCI,EP_NOHSDP,EP_AGE65,EP_AGE17,...,EP_MOBILE,EP_CROWD,EP_NOVEH,EP_GROUPQ,SVI,V_REPORT_COMPLETENESS,V_PCT_ALL,V_PCT_12YRS,V_PCT_18YRS,V_PCT_65YRS
1928,PA,Montour,42093,18294,11.1,2.2,34095,9.3,20.4,20.1,...,7.2,1.7,11.8,4.1,Low-Mod,94.4,70.4,81.6,83.9,98.7
971,PA,Lehigh,42077,362613,12.5,6.5,32252,11.4,16.2,22.8,...,2.1,2.0,8.9,2.5,Mod-High,94.4,64.7,75.7,78.2,99.9
1923,PA,Lackawanna,42069,211454,15.1,4.8,28427,9.1,19.3,20.3,...,2.6,1.5,10.1,3.8,Mod-High,94.4,65.5,75.5,77.3,94.7
2355,PA,Delaware,42045,563527,10.0,6.0,38251,7.2,15.7,22.2,...,0.3,1.5,10.6,3.9,Low-Mod,94.4,64.3,75.1,75.9,93.4
952,PA,Bucks,42017,626370,6.1,4.5,44103,6.1,17.6,20.8,...,1.9,1.1,4.8,1.5,Low,94.4,63.7,72.8,73.7,86.6


In [37]:
df_pa_order.tail(5)

,STATE,COUNTY,FIPS,E_TOTPOP,EP_POV,EP_UNEMP,EP_PCI,EP_NOHSDP,EP_AGE65,EP_AGE17,...,EP_MOBILE,EP_CROWD,EP_NOVEH,EP_GROUPQ,SVI,V_REPORT_COMPLETENESS,V_PCT_ALL,V_PCT_12YRS,V_PCT_18YRS,V_PCT_65YRS
951,PA,Bradford,42015,61304,12.0,4.8,27319,11.1,20.3,22.1,...,15.1,1.7,7.4,1.3,Low-Mod,94.4,37.5,43.7,45.9,63.6
973,PA,McKean,42083,41806,16.8,6.8,25517,8.9,18.5,20.0,...,8.7,0.9,9.4,7.2,Mod-High,94.4,34.8,39.8,41.9,57.7
949,PA,Bedford,42009,48611,12.7,4.9,25163,12.5,21.8,20.0,...,14.6,1.7,5.7,1.0,Low-Mod,94.4,34.0,38.7,41.1,63.7
965,PA,Fulton,42057,14506,11.4,5.6,26082,12.7,20.6,20.3,...,18.1,0.9,6.5,0.2,Low,94.4,29.9,34.3,36.5,59.4
979,PA,Potter,42105,16937,14.3,6.6,24087,11.9,22.8,21.0,...,13.9,0.8,7.2,1.6,Low-Mod,94.4,29.7,34.0,35.9,52.4


#### Answer:

#### 5 counties in PA with the highest V_PCT_12YRS: 
- Montour
- Lehigh
- Lackawanna
- Delaware
- Bucks

#### 5 counties in PA with the lowest V_PCT_12YRS: 
- Bradford
- McKean
- Bedford
- Fulton
- Potter

### PROBLEM 2 (45 points)

In this problem I ask you to implement in Python a simulation model of a simpified airline revenue management problem.

Consider an airline operating a flight with an aircraft whose capacity is 189 passenger seats. All seats on this aircraft are identical (that is, there is no business or first-class compartments). The marginal cost of each additional passenger is negligible. 

There are two fares for this flights: Standard and Premium. Standard tickets cost 249 dollars and have advance-purchase and other restrictions. Premium tickets cost 349 dollars and have far fewer restrictions.

Because of the advance-purchase restriction of the Standard tickets and the flexibility of the Premium tickets we will assume that **all** of the demand for Standard tickets materializes before any of the demand for Premium tickets. In other words, Standard fare passengers book early and Premium fare passengers book later.

It is estimated that the demand for Standard tickets follows approximately a Normal distribution with mean equal to 150 and a standard deviation equal to 30. The demand for Premium tickets is estimated to also follow approximately a Normal distribution but with mean equal to 100 and a standard deviation equal to 40. Assume the demands to be uncorrelated. (Do not worry about fractionality or the possibility of negative demand values).

For each part of this problem use 10,000 trials per simulation and a value of 123 for the random seed.

#### Part (a) 

Suppose the airline accepts all bookings, subject to capacity, on a first-come first-serve basis. Run a simulation to compute the expected (or average) revenue per flight?

### Approach: 

* In part (a), as there is no limit set on the number of standard tickets that can be purchased, we will use the random number generator normal function to determine the demand for the standard fare. The cap of 189 is applied to make sure airline accepts all bookings, subject to aircraft capacity. Using the determined standard fare demand, we calculate the 'standard revenue'.

* Next, as we assume that all of the demand for Standard tickets materializes before any of the demand for Premium tickets, based on the remaining number of seats in the aircraft after allocating standard fare seats, we run the simulation for the premium demand until the aircraft capacity is reached and calculate the corresponding 'premium revenue'.

* Total revenue is a summation of 'standard revenue' and 'premium revenue'.

* After simulating the above for 10000 trials, we take the average to determine the average revenue per flight.

In [1]:
# Importing required libraries/packages
import numpy as np

In [2]:
# Defining required variables

aircraft_capacity = 189
standard_ticket_price = 249
premium_ticket_price = 349
std_ticket_mean = 150
std_ticket_stdev = 30
premium_ticket_mean = 100
premium_ticket_stdev = 40
trials = 10000

In [3]:
tot_revenue =[]
rng = np.random.default_rng(seed=123)
for i in range(trials):
    
    standard_ticket_demand = rng.normal(loc= std_ticket_mean, scale = std_ticket_stdev)
    if standard_ticket_demand >= aircraft_capacity:
        total_seats_filled = aircraft_capacity
    else:
        total_seats_filled = int(standard_ticket_demand)
       
    revenue_standard = total_seats_filled * standard_ticket_price
       
    revenue_premium = 0
    if total_seats_filled < aircraft_capacity:
        premium_ticket_demand = rng.normal(loc = premium_ticket_mean, scale = premium_ticket_stdev)
        for demand in range(int(premium_ticket_demand)):
            revenue_premium += premium_ticket_price
            total_seats_filled += 1
            if total_seats_filled == aircraft_capacity:
                break
        total_revenue = revenue_standard + revenue_premium
        tot_revenue.append(total_revenue)
    else:
        total_revenue = revenue_standard + revenue_premium
        tot_revenue.append(total_revenue)

avg_tot_revenue = np.mean(tot_revenue)
print("The average revenue per flight is:", avg_tot_revenue)



The average revenue per flight is: 50176.0616


#### Part (b) 

Accepting customers on a first-come first-serve effectively dilutes revenue. Passengers requesting Standard tickets books early and this limits (due to aircraft capacity) the airline's ability to issue higher-fare Premium tickets for passegers booking later. Therefore, the airline's revenue management system will implement a **booking limit** on Standard tickets. That is, it will set a limit on the maximum number of Standard tickets that will be issued on any given flight (even before it knows precisely the demand for Premium tickets). Requests for Standard ticket bookings declined due to the booking limit is lost demand.  

Run a simulation to compute the expected (or average) revenue per flight if the airline chooses to set a booking limit of 110 seats.

### Approach:

* The approach is similar to part(a) except for the fact that the booking limit of 110 is set on the max number of tickets that can be purchased using the standard fare per flight.

In [4]:
booking_limit = 110

In [5]:
tot_revenue =[]
rng = np.random.default_rng(seed=123)
for i in range(trials):
    
    standard_ticket_demand = rng.normal(loc= std_ticket_mean, scale = std_ticket_stdev)
    if standard_ticket_demand >= booking_limit:
        total_seats_filled = booking_limit
    else:
        total_seats_filled = int(standard_ticket_demand)
       
    revenue_standard = total_seats_filled * standard_ticket_price 
    
    revenue_premium = 0
    premium_ticket_demand = rng.normal(loc = premium_ticket_mean, scale = premium_ticket_stdev)
    for demand in range(int(premium_ticket_demand)):
        revenue_premium += premium_ticket_price
        total_seats_filled += 1
        if total_seats_filled == aircraft_capacity:
            break
    total_revenue = revenue_standard + revenue_premium
    tot_revenue.append(total_revenue)

    
avg_tot_revenue = np.mean(tot_revenue)
print("The average revenue per flight is:", avg_tot_revenue)


The average revenue per flight is: 52254.0921


### Conclusion:

* As expected by the airline's revenue management system, the average revenue per flight is going to be definitely higher if there is a booking limit set of the standard fare seats.

---
### END